File: leetcode/leetcode_687.js
const longestUnivaluePath = (root) => {
  let longest = 0;

  const dfs = (node) => {
    if (!node) {
      return 0;
    }
    let [leftCnt, rightCnt] = [dfs(node.left), dfs(node.right)];
    if (!!node.left && node.val === node.left.val) {
      leftCnt++;
    } else {
      leftCnt = 0;
    }
    if (!!node.right && node.val === node.right.val) {
      rightCnt++;
    } else {
      rightCnt = 0;
    }
    longest = Math.max(longest, leftCnt + rightCnt);
    return Math.max(leftCnt, rightCnt);
  };
  dfs(root);
  return longest;
};

File: leetcode/leetcode_687.py
class Solution:
    def longestUnivaluePath(self, root):
        self.longest = 0
        def dfs(node):
            if not node: return 0

            leftCnt, rightCnt = dfs(node.left), dfs(node.right)
            if node.left and node.val == node.left.val:
                leftCnt += 1
            else:
                leftCnt = 0
            if node.right and node.val == node.right.val:
                rightCnt += 1
            else:
                rightCnt = 0
            self.longest = max(self.longest, leftCnt + rightCnt)
            return max(leftCnt, rightCnt)
            
        dfs(root)
        return self.longest